aten/src/ATen/ConjugateFallback.cpp

```diff
@@ -0,0 +1,32 @@
+#include <ATen/ConjugateFallback.h>
+#include <ATen/native/UnaryOps.h>
+
+namespace at {
+    void conjugateFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+        // Unwrap all arguments
+        const auto num_arguments = op.schema().arguments().size();
+        const auto arguments = torch::jit::last(stack, num_arguments);
+        auto args = torch::jit::pop(*stack, num_arguments);
+
+        for (int64_t idx = 0; idx < arguments.size(); ++idx) {
+            const auto& ivalue = arguments[idx];
+            if (!ivalue.isTensor()) {
+                continue;
+            }
+            auto* impl = args[idx].unsafeToTensorImpl();
+            if (impl->key_set().has(DispatchKey::Conjugate)) {
+                const auto& tensor = ivalue.toTensor();
+                // TODO: I think I want to take the existing tensor on the stack
+                // and run conjugation on it.
+                // The existing conj() function creates a new Tensor.
+                // Is it reaonsable to set the memory on the stack in place like this?
+                // I don't see other examples of fallbacks doing this,
+                // so instead I'm pushing the new conjugated vector onto the stack
+                auto conjugated_tensor = tensor.conj();
```

Comment:
See the comment here- I don't think I Have a clear understanding how a given fallback in the dispatch workflow is supposed to modify the stack.

My understanding is that the stack represents the call stack, i.e. the top of it contains the arguments to the kernel function that will eventually get called after all other dispatch fallback functions.

Should my conjugate fallback be modifying those argument in place? Pushing new arguments onto the stack?

Reply:
The stack is a mutable data structure, and the invariants are that when you call some downstream function, the top N arguments of the stack (however many arguments the function takes) are the arguments. On return, the invariant is that the input arguments had been popped from the stack, and the outputs are pushed onto the stack.

```
before stack: [arg1, arg2, arg3]  # NB: maybe it's the other way, I can never remember
call function
after stack: [ret1, ret2]
```

As the stack is mutable, an easy way you can modify the stack is simply by overwriting the direct entries with the new, explicitly conjugated (not using the bit) version. I'm not sure if the stack API has a direct way of expressing this, but stacks are just vectors so you can implement it by hand yourself.

Reply:
Yeah, I looked over your implementation and it doesn't work. I can give you more hints about how to fix it, but perhaps the invariant described above is enough?

Reply:
BTW, the use of `conj()` here directly is also unlikely to work, because you still have the conjugate bit on the tensor and you'll redispatch here. You need an operation like `conj_materialize()` which will take a tensor whose conjugate is bit, and returns the conjugated version *without the conjugate bit set*. You could probably do this by first allocating a new tensor ala `conj_view` with the conjugate bit turned off, and then calling `conj()`

Reply:
thanks-  I fixed the issues described (I think), but I'm still hitting issues. It might clarify my understanding of the stack in relation to a dispatch chain:

A dispatch "chain" is a sequence of function calls, consisting of (possibly several) "pre-processing" functions followed by a single backend kernel function. The set of functions called are in order of the priority of the dispatch keys, and are determined by:

1) the set of dispatch keys in the current context (in the tensor arguments, and local/global context)
(2) the mappings from from a given op/dispatch key to kernel functions or "fallback" functions. In the case of this conjugate function, we want the function to be called for all tensors with the Conjugate dispatch key set, regardless of the op)

A reasonable dispatch chain in the case of my change: Conjugate -> Add(cpu).

Now, I'm currently thinking of the stack as being passed along to each of the functions in the dispatch chain, such that the state of the stack at the end of the first function is the same as the state at the beginning of the second function.

E.g. suppose: we have:
`Tensor a = ...
 Tensor b = ...
Tensor c = a.conj_view() + b`

Invoking the "add" function should eventually invoke the dispatcher, first calling conjugateFallback() and then the add() kernel. I'd expect the stack at the beginning of the call to conjugateFallback to look like:
`[a (tensor, conj bit set); b (tensor, conj bit not set)]`
and after the call to look like:
`[a (conj bit NOT set, but the imaginary part negated); b (unchanged, conj bit not set)]`
(What we'd want the stack to look like at the beginning of the call to the add() kernel)

Which I believe is what the above is now doing. I'm currently hitting an issue however, where the arguments currently at the top of the stack are not being recognized as tensors (their tag is set to `c10::IValue::Tag::None`). Later, the call to the add() kernel function fails with "There were no tensor arguments to this function (e.g., you passed an empty list of Tensors)", which makes me think I corrupted the stack somewhere in my conjugate fallback.

Reply:
This description looks accurate.